src/neo4j_genai/retrievers/external/pinecone/pinecone.py

```diff
@@ -0,0 +1,157 @@
+from neo4j_genai.types import (
+    EmbedderModel,
+    Neo4jDriverModel,
+    PineconeClientModel,
+    PineconeNeo4jRetrieverModel,
+    PineconeSearchModel,
+    RawSearchResult,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class PineconeNeo4jRetriever(ExternalRetriever):
+    def __init__(
+        self,
+        driver: neo4j.Driver,
+        client: Pinecone,
+        index_name: str,
+        id_property_neo4j: str,
+        embedder: Optional[Embedder] = None,
+        return_properties: Optional[list[str]] = None,
+        retrieval_query: Optional[str] = None,
+        format_record_function: Optional[Callable] = None,
+    ):
+        try:
+            driver_model = Neo4jDriverModel(driver=driver)
+            client_model = PineconeClientModel(client=client)
+            embedder_model = EmbedderModel(embedder=embedder) if embedder else None
+            validated_data = PineconeNeo4jRetrieverModel(
+                driver_model=driver_model,
+                client_model=client_model,
+                index_name=index_name,
+                id_property_neo4j=id_property_neo4j,
+                embedder_model=embedder_model,
+                return_properties=return_properties,
+                retrieval_query=retrieval_query,
+                format_record_function=format_record_function,
+            )
+        except ValidationError as e:
+            raise ValueError(f"Validation failed: {e.errors()}")
+
+        super().__init__(
+            driver=driver,
+            id_property_external="id",
+            id_property_neo4j=validated_data.id_property_neo4j,
+        )
+        self.driver = validated_data.driver_model.driver
+        self.client = validated_data.client_model.client
+        self.index_name = validated_data.index_name
+        self.index = self.client.Index(index_name)
+        self.embedder = (
+            validated_data.embedder_model.embedder
+            if validated_data.embedder_model
+            else None
+        )
+        self.return_properties = validated_data.return_properties
+        self.retrieval_query = validated_data.retrieval_query
+
+    def _get_search_results(
+        self,
+        query_vector: Optional[list[float]] = None,
+        query_text: Optional[str] = None,
+        top_k: int = 5,
+        pinecone_filter: Optional[dict[str, Any]] = None,
+    ) -> RawSearchResult:
+        """Get the top_k nearest neighbor embeddings using Pinecone for either provided query_vector or query_text.
+        Both query_vector and query_text can be provided.
+        If query_vector is provided, then it will be preferred over the embedded query_text
+        for the vector search.
+        If query_text is provided, then it will check if an embedder is provided and use it to generate the query_vector.
+
+        See the following documentation for more details:
+        - [Query a vector index](https://neo4j.com/docs/cypher-manual/current/indexes-for-vector-search/#indexes-vector-query)
```

Comment:
We should change the links to reStructuredText format and add an example like in this PR: https://github.com/neo4j/neo4j-genai-python/pull/47/files#diff-1193271066bca242345f8030a71f70a53c3014ebae4b3186ca6574ed94654a1cR97-R120

Reply:
Done!